src/exchange/order_handler.cairo

```diff
@@ -177,50 +214,202 @@ mod OrderHandler {
+            // Check only order keeper.
+            let role_module_state = RoleModule::unsafe_new_contract_state();
+            role_module_state.only_order_keeper();
+
+            // Fetch data store.
+            let base_order_handler_state = BaseOrderHandler::unsafe_new_contract_state();
+            let data_store = base_order_handler_state.data_store.read();
+
+            global_reentrancy_guard::non_reentrant_before(data_store);
+            oracle_modules::with_oracle_prices_before(
+                base_order_handler_state.oracle.read(),
+                data_store,
+                base_order_handler_state.event_emitter.read(),
+                oracle_params.clone()
+            );
+
+            // TODO: Did not implement starting gas and try / catch logic as not available in Cairo
+            self._execute_order(key, oracle_params, get_contract_address());
+
+            oracle_modules::with_oracle_prices_after();
+            global_reentrancy_guard::non_reentrant_after(data_store);
         }
 
         fn simulate_execute_order(
             ref self: ContractState, key: felt252, params: SimulatePricesParams
-        ) { // TODO
-        }
+        ) {
+            // Check only order keeper.
+            let role_module_state = RoleModule::unsafe_new_contract_state();
+            role_module_state.only_order_keeper();
 
-        fn execute_order_keeper(
-            ref self: ContractState,
-            key: felt252,
-            oracle_params: SetPricesParams,
-            keeper: ContractAddress
-        ) { // TODO
+            // Fetch data store.
+            let base_order_handler_state = BaseOrderHandler::unsafe_new_contract_state();
+            let data_store = base_order_handler_state.data_store.read();
+
+            global_reentrancy_guard::non_reentrant_before(data_store);
+            oracle_modules::with_simulated_oracle_prices_before(
+                base_order_handler_state.oracle.read(), params.clone()
+            );
+
+            let oracle_params: SetPricesParams = Default::default();
+            self._execute_order(key, oracle_params, get_contract_address());
+
+            oracle_modules::with_simulated_oracle_prices_after();
+            global_reentrancy_guard::non_reentrant_after(data_store);
         }
     }
 
-    // *************************************************************************
+    // ***********************************************a**************************
     //                          INTERNAL FUNCTIONS
     // *************************************************************************
     #[generate_trait]
     impl InternalImpl of InternalTrait {
+        /// Executes an order.
+        /// # Arguments
+        /// * `key` - The key of the order to execute.
+        /// * `oracle_params` - The oracle params to set prices before execution.
+        /// * `keeper` - The keeper executing the order.
+        fn _execute_order(
```

Comment:
This function is marked as external in the solidity code

Reply:
I think there are two functions - one external and another internal one (see line 311 vs 369)

Reply:
Yeah but if I'm not mistaken they're both marked as external in solidity even though one is prefixed with an _